<a href="https://colab.research.google.com/github/josefinabernardo/project-euler/blob/main/Project_Euler_Part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Project Euler - Part 1
This notebook includes my solutions to problems 2, 8, 13, and 22 of the Project Euler competition held on HackerRank. I include a short description of each problem and demonstrate how the code works on the sample inputs from HackerRank.

In [ ]:
import sys
import math

## Problem 2: Even Fibonacci numbers
By considering the terms in the Fibonacci sequence whose values do not exceed **N**, find the sum of the even-valued terms.

In [ ]:
# Problem 2: Even Fibonacci numbers
t = int(input().strip())
for a0 in range(t):
    n = int(input().strip())
    seq = [1,2,3,5,8]
    while seq[-1] + seq[-2] <= n:
        seq.append(seq[-1]+seq[-2])
    seqlist = [i for i in seq if i%2 == 0]
    print(sum(seqlist))

2
10
10
100
44


**Explanation:** We specify the number of input lines (2) and then input an N (here 10 and 100) on each line. The code outputs the sum of even Fibonacci numbers below N (2 + 8 = 10, 2 + 8 + 34 = 44).

##Problem 8: Largest product in a series
Find the greatest product of **K** consecutive digits in the **N** digit number.

In [ ]:
# Problem 8: Largest product in a series
t = int(input().strip())
for a0 in range(t):
    n,k = input().strip().split(' ')
    n,k = [int(n),int(k)]
    num = input().strip()
    num = [int(d) for d in str(num)]
    print(max([math.prod(num[s:e]) for s,e in zip(range(0,n-k+1),range(k,n+1))]))

2
10 5
3675356291
3150
10 5
2709360626
0


**Explanation:** We specify the number of input lines (2) and then input an N, a K, and the number to calculate the product for. The code outputs the products. For the second number, the maximum product is 0 because each possible consecutive 5 numbers contain 0.

##Problem 13: Large sum
Work out the first ten digits of the sum of **N 50-digit** numbers.

In [ ]:
# Problem 13: Large sum
n = int(input().strip())
my_sum = 0
for a0 in range(n):
    num = int(input().strip())
    my_sum += num
print(int(str(my_sum)[0:10]))

5
37107287533902102798797998220837590246510135740250
46376937677490009712648124896970078050417018260538
74324986199524741059474233309513058123726617309629
91942213363574161572522430563301811072406154908250
23067588207539346171171980310421047513778063246676
2728190129


**Explanation:** We specify the number of input lines (5) and then input the 50 digit numbers. The code outputs the first 10 digits of their sum.

##Problem 22: Names scores
You are given around five-thousand first names, begin by sorting it into alphabetical order. Then working out the alphabetical value for each name, multiply this value by its alphabetical position in the list to obtain a name score.

In [ ]:
# Problem 22: Names scores
n = int(input().strip())
name_list = []
for a0 in range(n):
    name = input().strip()
    name_list.append(name)
q = int(input().strip())
for b0 in range(q):
    query = str(input().strip())
    print(sum([ord(l)-ord("A")+1 for l in query])*(sorted(name_list).index(query)+1))

5
ALEX
LUIS
JAMES
BRIAN
PAMELA
1
PAMELA
240


**Explanation:** First, we specify the number of names (5) and then input the names. Second, we specify the number of queries (1) and the name(s) to query. The code multiplies the alphabetical value of the queried name (e.g. PAMELA becomes 16+1+13+5+12+1 = 48) with its position. After sorting the list alphabetically, Pamela is still the fifth name on the list so by multiplying we obtain 240 (5*48 = 240).